### DISCUSSION TITLE
ETL webhooks

### DISCUSSION THREAD

**Discussion by djordjedjukic** [ravendb/ravendb#12199] [Ideas]
Do you maybe have in plan to make some kind of webhook like ETL for ravendb or SQL?

_Example:_
I want to send an email when a new order comes in. So I want to define some API endpoint in ETL which should be called when a new order is created, and then I can do the work with the data endpoint received.

**Comment by ayende** [comment]
What would the interface look like?

We can use the script to transform the document to a payload to a URL, but there are many issues we have to deal with.

* Support `GET` vs `POST`, etc.
* Authentication? - Webhooks are generally without it, no?
* How do you handle retries? Normally, I believe would be fine.

In general, that should be fairly similar to the rest of our ETL providers, and can certainly generate some good usability to users.

**Reply by djordjedjukic** [reply]
I guess that authentication could be done with API key.

For me personally, it would be very useful, because I could use raven like a queue in most cases. With this, I would avoid some schedulers for checking is there something new to process.

**Reply by ayende** [reply]
Authentication via API key is something that we need to handle, yes. It is sometimes handled via a query string, but often via a header, so that would be needed.
The API we offer should probably mock the `jQuery` one, I guess? 

As an aside, we are planning integration with RabbitMQ, Kafka, etc. Might serve the same purpose.

**Reply by genachka** [reply]
Am I wrong in thinking that this seems kind of like a trigger / event if yes, that would be very helpful for us too. For our use cases, standing up a traditional message queue system isn't ideal and can be more costly, but we leverage things like Pubnub and sometimes AWS SNS. At the database side if we could write in Javascript for example a trigger (both Pubnub and AWS SNS have Javascript SDKs) that would trigger say when a document of metadata collection type `order` was created it would fire and publish a message on the topic defined. Short of hardcoding the authentication in the Javascript (not ideal of course) if there were a database `secrets` store much like GitHub allows for GitHub actions that would work well I think. And doing things like triggering a POST webhook to for example Slack for same kind of use case.

**Reply by ayende** [reply]
The 5.4 release currently supports both ETL to Kafka and ETL to RabbitMQ. That would probably be the easiest way to handle this.
Supporting `POST` requests and their mechanics is on teh table, but will likely not be in the next release, so quie some time ni the future.

**Reply by genachka** [reply]
@ayende thanks I did see 5.4 released with that. What's the timeline to migrate cloud to 5.4 or can you point me to a document that details this?

While in general I'm a fan of and have used both Kafka and RabbitMQ in past, at this time unfortunately that would require standing up additional resources. We're running on RavenDB cloud and our apps are leveraging microservices and serverless designs so introducing additional dedicated compute for the message broker / queue isn't feasible for our use cases. 

A native RavenDB trigger that we could execute code (I bring up Javascript as it seems to be an acceptable approach for Map Reduce as well already) that doesn't require standing up another dedicated message broker / queue of course would be more ideal for us but that sounds like may not be something in the cards anytime soon if ever.

We already leverage AWS SNS so that would also work if was a future addition -- but if there's concern for being tied to AWS, a webhook would suffice since it allows being leveraged by so many endpoints and we can also leverage AWS API Gateway that will trigger a Lambda function that could then publish a message to AWS SNS if needed and still be all serverless.

**Reply by ayende** [reply]
The 5.4 is available _now_ on the cloud. You can request upgrade to it via a support request.

About `POST` / `SNS` support - both are on the roadmap, but the schedule for the next release is full, so it will be later than that.

Of course, we would be *very* happy to accept any pull requests.
